crates/server/src/server/server_future.rs

```diff
@@ -1045,49 +1091,86 @@ pub(crate) async fn handle_request<R: ResponseHandler, T: RequestHandler>(
         request_handler.handle_request(&request, reporter).await;
     };
 
+    // method to return an error to the client
+    let error_response_handler = |protocol: Protocol,
+                                  src_addr: SocketAddr,
+                                  header: Header,
+                                  query: LowerQuery,
+                                  response_code: ResponseCode,
+                                  error: Box<ProtoError>,
+                                  response_handler: R| async move {
+        // debug for more info on why the message parsing failed
+        debug!(
+            "request:{id} src:{proto}://{addr}#{port} type:{message_type} {op}:{response_code}:{error}",
+            id = header.id(),
+            proto = protocol,
+            addr = src_addr.ip(),
+            port = src_addr.port(),
+            message_type = header.message_type(),
+            op = header.op_code(),
+            response_code = response_code,
+            error = error,
+        );
+
+        // The reporter will handle making sure to log the result of the request
+        let mut reporter = ReportingResponseHandler {
+            request_header: header,
+            query,
+            protocol,
+            src_addr,
+            handler: response_handler,
+        };
+
+        let response = MessageResponseBuilder::new(None);
+        let result = reporter
+            .send_response(response.error_msg(&header, response_code))
+            .await;
+
+        if let Err(e) = result {
+            warn!("failed to return FormError to client: {}", e);
+        }
+    };
```

Comment:
It's not obvious why this was moved into a closure?

Reply:
See below, the calls to `error_response_handler` now occur in to of the match arms.

Reply:
see the usage across the two match arms below.